stl/inc/ranges

```diff
@@ -9553,6 +9553,604 @@ namespace ranges {
+            }
+
+            template <size_t _Index = sizeof...(_Rest)>
+            constexpr void _Advance(const _Difference_type<_Const> _Off) {
+                if (_Off == 0) {
+                    return;
+                }
+
+                auto& _Range = _STD get<_Index>(_Parent->_Bases);
+                auto& _It    = _STD get<_Index>(_Current);
+                using _Iter  = remove_reference_t<decltype(_It)>;
+                using _Diff  = _Difference_type<_Const>;
+
+                if constexpr (_Index > 0) {
+                    const auto _Size       = static_cast<_Diff>(_RANGES ssize(_Range));
+                    const auto _Begin      = _RANGES begin(_Range);
+                    const auto _It_off     = static_cast<_Diff>(_It - _Begin);
+                    const auto _It_new_off = static_cast<_Diff>((_It_off + _Off) % _Size);
+                    const auto _Next_off   = static_cast<_Diff>((_It_off + _Off) / _Size);
+                    if (_It_new_off < 0) {
+                        _It = _Begin + static_cast<iter_difference_t<_Iter>>(_It_new_off + _Size);
+                        _Advance<_Index - 1>(_Next_off - 1);
+                    } else {
+                        _It = _Begin + static_cast<iter_difference_t<_Iter>>(_It_new_off);
+                        _Advance<_Index - 1>(_Next_off);
+                    }
+                } else {
+#if _ITERATOR_DEBUG_LEVEL != 0
+                    const auto _It_off = static_cast<_Diff>(_It - _RANGES begin(_Range));
+                    _STL_VERIFY(_It_off + _Off >= 0, "Cannot advance cartesian_product_view iterator before begin "
+                                                     "(N4928 [range.cartesian.iterator]/19).");
+                    if constexpr (sized_range<decltype(_Range)>) {
+                        const auto _Size = static_cast<_Diff>(_RANGES ssize(_Range));
+                        _STL_VERIFY(_It_off + _Off < _Size
+                                        || (_It_off + _Off == _Size
+                                            && _Entire_tail_at_begin(make_index_sequence<sizeof...(_Rest)>{})),
+                            "Cannot advance cartesian_product_view iterator past end (N4928 "
+                            "[range.cartesian.iterator]/19).");
+                    }
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+                    _It += static_cast<iter_difference_t<_Iter>>(_Off);
+                }
+            }
+
+            template <size_t... _Indices>
+            _NODISCARD constexpr bool _Is_end(index_sequence<_Indices...>) const {
+                return ((_STD get<_Indices>(_Current) == _RANGES end(_STD get<_Indices>(_Parent->_Bases))) || ...);
+            }
+
+            template <size_t... _Indices>
+            _NODISCARD constexpr auto _End_tuple(index_sequence<_Indices...>) const {
+                return tuple{_RANGES end(_STD get<0>(_Parent->_Bases)),
+                    _RANGES begin(_STD get<_Indices + 1>(_Parent->_Bases))...};

```

Comment:
This CTAD might break in a pathological case, namely when the result of `end` is a `std::pair` that provides everything required by `input_iterator`.

IMO it's safer to write out the template arguments in generic code.

Occurs below in `cartesian_product_view::end`.